### PR TITLE
feat(core): list azure tenants [BLOCKED]

### PR DESCRIPTION
**Pull request by deep-cognite**
Most of this code is copy-paste of the methods used for AAD CDF auth stuff. 

https://docs.microsoft.com/en-us/rest/api/resources/tenants/list

--

Status: BLOCKED (23.08): ref: https://cognitedata.slack.com/archives/C6KNJCEEA/p1629717852108600

### CODE REVIEW

File: packages/core/src/baseCogniteClient.ts [hunk 1]
```diff
@@ -911,18 +921,72 @@ export default class BaseCogniteClient {
...
+      },
+    };
+    const azureAdClient = new AzureAD({ config, cluster: '', debug });
+
+    let accessToken = null;
+
+    const account = this.validateAccount(azureAdClient);
+    if (account) {
+      accessToken = await azureAdClient.getAzureToken();
+    }
```
Comment (deep-cognite) [inline packages/core/src/baseCogniteClient.ts]: No idea if this is needed, followed the AD flow above.

File: packages/core/src/baseCogniteClient.ts [hunk 2]
```diff
@@ -911,18 +921,72 @@ export default class BaseCogniteClient {
...
+  }
+
+  /**
+   * Disclaimer: This only works with AAD.
+   *
+   * Fetches the current users tenant list (https://docs.microsoft.com/en-us/rest/api/resources/tenants/list#code-try-0)
+   * @param clientId
+   * @param debug
+   */
+  public async getTenantList(
```
Comment (deep-cognite) [inline packages/core/src/baseCogniteClient.ts]: rename to
```suggestion
  public async getAzureTenantList(
```
...?

File: packages/core/src/baseCogniteClient.ts [hunk 3]
```diff
@@ -911,18 +921,72 @@ export default class BaseCogniteClient {
...
+      accessToken = await azureAdClient.getAzureToken();
+    }
+
+    if (!accessToken) {
+      return [];
+    }
+
+    const http = new CDFHttpClient(
+      'https://management.azure.com/tenants?api-version=2020-01-01',
+      this.getRetryValidator()
```
Comment (deep-cognite) [inline packages/core/src/baseCogniteClient.ts:979]: Not sure if that should be included, here.

Comment (deep-cognite) [reply packages/core/src/baseCogniteClient.ts:979]: Used `CDFHttpClient` since that was nicely setup for http requests.

**Comment by ngerstle-cognite** [comment]
@deep-cognite 
Does this require `user_impersonation` for the Azure AD management api?
What does the consent prompt specify?

(I'm not sure we can do this if so- I don't think the customers would be happy to issue tokens with potential admin access to their azure subscription/ad just to list tenants..)

**Comment by deep-cognite** [comment]
> @deep-cognite
> Does this require `user_impersonation` for the Azure AD management api?
> What does the consent prompt specify?
> 
> (I'm not sure we can do this if so- I don't think the customers would be happy to issue tokens with potential admin access to their azure subscription/ad just to list tenants..)

Yeah, we need `user_impersonation` in order to get the tenants list: https://docs.microsoft.com/en-us/rest/api/resources/providers/list-at-tenant-scope#scopes

Not sure about the consent window, forgot to take a screenshot and I don't know how to get it back now.

I did some research, and there are more people that have the same concern, @ngerstle-cognite: https://docs.microsoft.com/en-us/answers/questions/442040/list-tenants-for-current-user-without-user-imperso.html
^ This was newly posted, as well. Hopefully the MS Graph team can change the scope to User.Read or something 🙈

**Comment by jarlah** [comment]
What about auth0? Other custom idps? I dont think this should live in the js sdk

**Comment by jarlah** [comment]
We can create a new package that extends the functionality maybe. Extends the base client.

**Comment by ngerstle-cognite** [comment]
> > @deep-cognite
> > Does this require `user_impersonation` for the Azure AD management api?
> > What does the consent prompt specify?
> > (I'm not sure we can do this if so- I don't think the customers would be happy to issue tokens with potential admin access to their azure subscription/ad just to list tenants..)
> 
> Yeah, we need `user_impersonation` in order to get the tenants list: https://docs.microsoft.com/en-us/rest/api/resources/providers/list-at-tenant-scope#scopes
> 
> Not sure about the consent window, forgot to take a screenshot and I don't know how to get it back now.
> 
> I did some research, and there are more people that have the same concern, @ngerstle-cognite: https://docs.microsoft.com/en-us/answers/questions/442040/list-tenants-for-current-user-without-user-imperso.html
> ^ This was newly posted, as well. Hopefully the MS Graph team can change the scope to User.Read or something see_no_evil

If this functionality requires that scope, it's a no-go for most customers (I wouldn't be comfortable granting that as an AAD admin myself).

**Comment by ngerstle-cognite** [comment]
> What about auth0? Other custom idps? I dont think this should live in the js sdk

I do think there's a valid question around specialization. To this point, we've stated that we'll support AAD, and other IdPs meeting the requirements, and that we'll only provide guides/tutorials/native support for AAD. I think it's reasonable to add AAD specific code for retrieving a token to the SDK, as long as it's not required (an alternative call to use another IdP).
It would be fantastic to have a solid OAuth2/OIDC client that's standards oriented, and not AAD specific.


I'd also point out that this functionality is relevant for user sign in, but this SDK may be used for other clients (this is a specific request- how to use the JS sdk with backend apis/scripts/etc)- in that sense, a different package/product might be best (no one else has made a lib to handle this already?).

**Comment by rek** [comment]
> What about auth0? Other custom idps? I dont think this should live in the js sdk
this sdk is full of Azure+AAD things. when there are requirements for other idps... someone can do the work to support them.

> but this SDK may be used for other clients 
again yes it might, but failing having a team (or even a person) to support this sdk, we can only contribute the code we have a requirement for

this part of the sdk already deals with a lot of similar things, adding this extra method here seems very inline with the current methodology here.

File: packages/core/src/authFlows/aad.ts [hunk 4]
```diff
@@ -186,6 +207,10 @@ export class AzureAD {
     ];
   }
 
+  private getAzureScopes(): string[] {
```
Comment (rek) [inline packages/core/src/authFlows/aad.ts:210]:
```suggestion
  private getAzureManagementScopes(): string[] {
```

File: packages/core/src/authFlows/aad.ts [hunk 5]
```diff
@@ -163,6 +171,19 @@ export class AzureAD {
...
   }
 
+  async getAzureToken(): Promise<string | null> {
+    if (!this.getAccount()) return null;
+
+    const {
+      accessToken,
+    }: AuthenticationResult = await this.msalApplication.acquireTokenSilent({
+      account: this.getAccount(),
+      scopes: this.getAzureScopes(),
```
Comment (rek) [inline packages/core/src/authFlows/aad.ts:181]:
```suggestion
      scopes: this.getAzureManagementScopes(),
```

File: packages/core/src/authFlows/aad.ts [hunk 6]
```diff
@@ -163,6 +171,19 @@ export class AzureAD {
     return accessToken;
   }
 
+  async getAzureToken(): Promise<string | null> {
```
Comment (rek) [inline packages/core/src/authFlows/aad.ts:174]:
```suggestion
  async getAzureManagementToken(): Promise<string | null> {
```

**Pull request closed without merging**